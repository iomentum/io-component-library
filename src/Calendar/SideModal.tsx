import React, { Dispatch, memo, SetStateAction } from "react";
import { Button, Modal, Slide } from "@material-ui/core";
import { SideModalContent } from "./MyCalendar.style";

export enum OpeningDirection {
  Left = "left",
  Right = "right",
  Up = "up",
  Down = "down",
}

interface SideModalProps {
  openState: [boolean, Dispatch<SetStateAction<boolean>>];
  openingDirection: OpeningDirection;
  onSave: () => void;
  children: JSX.Element[];
}

export const SideModal = memo((props: SideModalProps) => {
  const [openModal, setOpenModal] = props.openState;

  const handleOnSave = () => {
    props.onSave();
    setOpenModal(false);
  }

  return (
    <Modal open={openModal} onClose={() => setOpenModal(false)}>
      <Slide direction={props.openingDirection} in={openModal}>
        <SideModalContent elevation={1}>
          {props.children}
          <Button
            variant="contained"
            color="primary"
            onClick={handleOnSave}
          >
            Save
          </Button>
        </SideModalContent>
      </Slide>
    </Modal>
  );
});
