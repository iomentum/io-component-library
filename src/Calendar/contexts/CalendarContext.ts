import { createContext, Dispatch, SetStateAction } from "react";
import { Display, MomentRangeExtended } from "../utils";

export interface CalendarContextInterface {
  display: Display;
  setDisplay: Dispatch<SetStateAction<Display>>;
  setDate: Dispatch<SetStateAction<MomentRangeExtended>>;
}

export const CalendarContext = createContext<CalendarContextInterface>(null);
