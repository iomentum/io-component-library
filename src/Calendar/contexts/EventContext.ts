import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, Event } from '../reducers/EventReducer';

export interface EventContextInterface {
  eventsCollection: Event[];
  setEventsCollection: Dispatch<SetStateAction<Event[]>>;
  eventManagementOpened: boolean;
  setEventManagementOpened: Dispatch<SetStateAction<boolean>>;
  event: Event;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventContext = createContext<EventContextInterface>(null);
