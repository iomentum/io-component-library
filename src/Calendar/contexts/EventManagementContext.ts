import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, EventModel } from '../reducers/EventReducer';

export interface EventManagementContext {
  fullDayEvent: boolean;
  event: EventModel;
  setFullDayEvent: Dispatch<SetStateAction<boolean>>;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventManagementContext = createContext<EventManagementContext>(null);
