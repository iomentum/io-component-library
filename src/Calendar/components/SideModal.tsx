import React, { useContext } from 'react';
import { Modal, Slide } from '@material-ui/core';
import { SideModalContent } from './SideModal.style';
import { EventContext } from '../contexts/EventContext';

interface SideModalProps {
  children: JSX.Element[];
}

export const SideModal = (props: SideModalProps) => {
  const { eventManagementOpened: openedModal, setEventManagementOpened: setOpenModal } =
    useContext(EventContext);

  return (
    <Modal open={openedModal} onClose={() => setOpenModal(false)}>
      <Slide direction="left" in={openedModal}>
        <SideModalContent elevation={1}>{props.children}</SideModalContent>
      </Slide>
    </Modal>
  );
};
