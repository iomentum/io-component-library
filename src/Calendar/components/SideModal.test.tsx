/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render, screen } from '@testing-library/react';
import { EventContext } from '../contexts/EventContext';
import { SideModal } from './SideModal';

const eventContextMock = (component, { providerValue, ...renderOptions }) =>
  render(
    <EventContext.Provider value={providerValue}>{component}</EventContext.Provider>,
    renderOptions
  );

const providerValue = (eventManagementOpened) => ({
  eventManagementOpened,
  setEventManagementOpened: jest.fn(),
});

describe('SideModal component', () => {
  describe('@snapshots', () => {
    it('should match with previous opened SideModal', () => {
      const { baseElement } = eventContextMock(
        <SideModal onSave={jest.fn()}>
          <div />
          <div />
        </SideModal>,
        {
          providerValue: providerValue(true),
        }
      );

      expect(baseElement).toMatchSnapshot('SideModal snapshot opened');
    });

    it('should match with previous closed SideModal', () => {
      const { baseElement } = eventContextMock(
        <SideModal onSave={jest.fn()}>
          <div />
          <div />
        </SideModal>,
        {
          providerValue: providerValue(false),
        }
      );

      expect(baseElement).toMatchSnapshot('SideModal snapshot closed');
    });
  });

  describe('@props', () => {
    it('should have children when opened', () => {
      eventContextMock(
        <SideModal onSave={jest.fn()}>
          <div data-testid="test" />
          <div />
        </SideModal>,
        {
          providerValue: providerValue(true),
        }
      );

      expect(screen.queryByTestId('test')).not.toBeNull();
    });

    it('should not have children when closed', () => {
      eventContextMock(
        <SideModal onSave={jest.fn()}>
          <div data-testid="test" />
          <div />
        </SideModal>,
        {
          providerValue: providerValue(false),
        }
      );

      expect(screen.queryByTestId('test')).toBeNull();
    });
  });

  describe('@events', () => {
    it('should handle save', () => {
      const onSave = jest.fn();
      const generatedProviderValue = providerValue(true);

      eventContextMock(
        <SideModal onSave={onSave}>
          <div data-testid="test" />
          <div />
        </SideModal>,
        {
          providerValue: generatedProviderValue,
        }
      );

      screen.getByRole('button').click();

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(generatedProviderValue.setEventManagementOpened).toHaveBeenCalledWith(false);
    });
  });
});
