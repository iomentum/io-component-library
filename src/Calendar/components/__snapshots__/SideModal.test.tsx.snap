// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideModal component @snapshots should match with previous closed SideModal: SideModal snapshot closed 1`] = `
<body
  style=""
>
  <div />
</body>
`;

exports[`SideModal component @snapshots should match with previous opened SideModal: SideModal snapshot opened 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: rgba(0, 0, 0, 0.5);"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root sc-bdnxRM eCTqnF MuiPaper-elevation1 MuiPaper-rounded"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div />
      <div />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Save
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
